app/registries/foreman/plugin/fact_parser_registry.rb

```diff
@@ -0,0 +1,24 @@
+module Foreman
+  class Plugin
+    class FactParserRegistry
+      attr_reader :parsers
+
+      def initialize
+        @parsers = {}.with_indifferent_access
+      end
+
+      def parser_for(type)
```

Comment:
Thoughts about using `[]` so the API becomes `Foreman::Plugin.fact_parser_registry[:puppet].new(facts)`. I think it feels confusing that you get a parser class with an API that's `parser_for`.

Another thing to consider is `parse(:puppet, facts)`.

Reply:
Good idea! After some thinking, I use a `[type]` call because it's support only one time get of parser, not like the second suggested option. I can clearly imagine that somewhere in code going to method like `parse_by_puppet_parser(facts)` in that case.